### PR TITLE
I/O Stack Simplification and Optimization

### PR DESCRIPTION
**Pull request by jbaiter**
The I/O stack in the plugin was previously very close to the way the `UnifiedHighlighter` in Solr worked. However, this was not really a good match, since the interfaces involved were highlgy stateful and with a very large surface area, which made implementations brittle and overly complicated. Additionally, the way we performed I/O incurred a lot of String construction, which proved to be a significant hotspot when profiling the plugin. This change set refactors the I/O stack to be much simpler. Additionally, performance has been improved a lot by significantly reducing the number of String allocations, which also led to a decrease in the number of filesystem reads.

Additionally, we no longer memory-map files, instead we simply use the regular `java.nio.FileChannel` API. This saves us one data copy on the JVM side, instead we let the OS fill the buffer for us. This nets us an additional performance improvement for multithreaded highlighting, while being roughly on par with mmapped IO in the single threaded version. An additional benefit is  that we now no longer [run the risk of crashing the JVM when an I/O error like a disappearing mount occurs](https://www.javacodegeeks.com/2014/03/detecting-write-failures-when-using-memory-mapped-files-in-java.html) 😅

The combination of these two changes make  the plugin a whole lot more ✨performant✨, with response times now all below the latency threshold for "sluggishness" (the orange line):

perfplot

As for the API simplifications, it boils down to the following:
- `IterableCharSequence` and its implementations are gone
- New Interface `SourceReader` with a base class `BaseSourceReader`, for which implementations only have to provide a `int read(byte[] dst, int dstOffset, int start, int len)` implementation

These changes should make it significantly easier to add support for new I/O backends, most importantly S3 (see #49).

### CODE REVIEW

File: src/main/java/com/github/dbmdz/solrocr/reader/BaseSourceReader.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    int startOffset = sectionIndex * sectionSize;
+    int readLen = Math.min(sectionSize, this.length() - startOffset);
+    this.readBytes(copyBuf, 0, startOffset, readLen);
+    // Construct a String without going through a decoder to save on CPU.
+    // Given that the method has been deprecated since Java 1.1 and was never removed, I don't think
+    // this is very risky 😅
+    Section section =
+        new Section(startOffset, startOffset + sectionSize, new String(copyBuf, 0, 0, readLen));
+    if (cache.length > 0 && cacheSlotsUsed == cache.length) {
```
Comment (schmika) [inline src/main/java/com/github/dbmdz/solrocr/reader/BaseSourceReader.java:226]: I think we need to check against `maxCacheEntries`:
```suggestion
    if (cache.length > 0 && cacheSlotsUsed == maxCacheEntries) {
```

File: src/main/java/com/github/dbmdz/solrocr/reader/BaseSourceReader.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ */
+public abstract class BaseSourceReader implements SourceReader {
+  private static final int UNUSED = -1;
+
+  protected final SourcePointer pointer;
+  protected final int sectionSize;
+  private final byte[] copyBuf;
+  private final int maxCacheEntries;
+
+  /** Array with a slot for every possible section in the source */
```
[the remark this answers is not in the document]
Comment (jbaiter) [reply src/main/java/com/github/dbmdz/solrocr/reader/BaseSourceReader.java:22]: Yes 👌🏾 Will do 👍🏾

File: src/main/java/com/github/dbmdz/solrocr/solr/SolrOcrHighlighter.java [hunk 3]
```diff
@@ -52,16 +52,21 @@
...
+  private final int readerMaxCacheEntries;
 
   public SolrOcrHighlighter() {
-    this(Runtime.getRuntime().availableProcessors(), 8);
+    this(Runtime.getRuntime().availableProcessors(), 8, 8 * 1024, 64 * 1024);
   }
 
-  public SolrOcrHighlighter(int numHlThreads, int maxQueuedPerThread) {
+  public SolrOcrHighlighter(
+      int numHlThreads, int maxQueuedPerThread, int readerSectionSize, int readerMaxCacheEntries) {
```
Comment (schmika) [inline src/main/java/com/github/dbmdz/solrocr/solr/SolrOcrHighlighter.java:64]: I think `readerMaxCacheEntries` is currently actually `maxSectionCacheSize`. Do we have to divide by `readerSectionSize` to get the maximum number of cache entries?

Comment (jbaiter) [reply src/main/java/com/github/dbmdz/solrocr/solr/SolrOcrHighlighter.java:64]: 😱 omg, yes 😱 thank you

File: docs/performance.md [hunk 4]
```diff
@@ -37,6 +37,18 @@ Generally speaking, local storage is better than remote storage (like NFS or CIF
...
+the OCR files in aligned sections and caches these to reduce the number of reads and allocations. The bigger
+the cache size, the more data is read from the disk, i.e. the chances of cache hits increase. However, this
+comes at the cost of more memory usage and more allocations in the JVM, which can have a performance impact.
+By default, the plugin uses a section size of 8KiB with a maximum number of cached sections of 10,
+which is a good trade-off for most setups and performed well in our benchmarks. If you want to tweak these
+settings, use the `sectionReadSizeKib` and `maxSectionCacheSizeKib` parameters on the `OcrHighlightComponent`
+in your `solrconfig.xml`:
+
+- `sectionReadSizeKib`: The size of the sections that are read from the OCR files. The default is 8KiB.
+- `maxSectionCacheSizeKib`: The maximum memory that is used for caching sections. The default is 10 * `sectionReadSizeKib`.
```
Comment (schmika) [inline docs/performance.md:50]: The default is currently 64KiB (see https://github.com/dbmdz/solr-ocrhighlighting/blob/c82c03f7494649f5fb701dc306561cc986f51e77/src/main/java/solrocr/OcrHighlightComponent.java#L95)
```suggestion
- `maxSectionCacheSizeKib`: The maximum memory that is used for caching sections. The default is 8 * `sectionReadSizeKib`.
```

**Review by schmika — APPROVED** [review]

**Pull request merged**